src/sync.cpp

```diff
@@ -36,6 +38,425 @@ const int Sync::FILE_UPDATE_DELAY_DS = 30;
+std::pair<size_t, bool> combinedFingerprint(FileSystemAccess& fsaccess, const set<string>& paths)
+{
+    size_t value = 0;
+    bool valid = false;
+    for (const auto& path : paths)
+    {
+        auto fa = std::unique_ptr<FileAccess>{fsaccess.newfileaccess()};
+        if (!fa->fopen(const_cast<string*>(&path), true, false))
+        {
+            LOG_err << "Unable to open path: " << path;
+            valid = false;
+            break;
+        }
+        if (fa->type == FILENODE)
+        {
+            FileFingerprint ffp;
+            ffp.genfingerprint(fa.get());
+            if (!ffp.isvalid)
+            {
+                LOG_err << "Invalid fingerprint: " << path;
+                valid = false;
+                break;
+            }
+            hashCombine(value, ffp.getHash());
+            valid = true;
+        }
+    }
+    return {value, valid};
+}
+
+// Represents a generalized fingerprint for file or folder, either contructed from a LocalNode or a filesystem path.
+class Fingerprint
+{
+public:
+    explicit Fingerprint(LocalNode& node)
+    {
+        if (node.type == FILENODE)
+        {
+            if (!node.isvalid)
+            {
+                LOG_err << "Invalid fingerprint: " << node.localname;
+                return;
+            }
+            mHash = node.getHash();
+            mIsValid = true;
+        }
+        else if (node.type == FOLDERNODE)
+        {
+            const auto res = combinedFingerprint(node.children);
+            mHash = res.first;
+            mIsValid = res.second;
+        }
+        else
+        {
+            assert(false && "Invalid node type");
+        }
+    }
+
+    explicit Fingerprint(FileSystemAccess& fsaccess, FileAccess& fa, const set<string>& paths)
+    {
+        if (fa.type == FILENODE)
+        {
+            FileFingerprint ffp;
+            ffp.genfingerprint(&fa);
+            if (!ffp.isvalid)
+            {
+                LOG_err << "Invalid fingerprint";
+                return;
+            }
+            mHash = ffp.getHash();
+            mIsValid = true;
+        }
+        else if (fa.type == FOLDERNODE)
+        {
+            const auto res = combinedFingerprint(fsaccess, paths);
+            mHash = res.first;
+            mIsValid = res.second;
+        }
+        else
+        {
+            assert(false && "Invalid node type");
+        }
+    }
+
+    MEGA_DISABLE_COPY(Fingerprint)
+    MEGA_DEFAULT_MOVE(Fingerprint)
+
+    size_t getHash() const
+    {
+        return mHash;
+    }
+
+    bool getIsValid() const
+    {
+        return mIsValid;
+    }
+
+    struct Hash
+    {
+        size_t operator()(const Fingerprint& node) const
+        {
+            assert(node.mIsValid);
+            return node.mHash;
+        }
+    };
+
+private:
+    size_t mHash = 0;
+    bool mIsValid = false;
+};
+
+bool operator==(const Fingerprint& lhs, const Fingerprint& rhs)
+{
+    return lhs.getHash() == rhs.getHash();
+}
```

Comment:
I think we can reuse FileFingerprint instead of this extra class - this one tries to boil a fingerprint down to a size_t but that will lose precision - FileFingerprint is already not perfectly guaranteed not to collide, but we should maintain that level of accuracy, I think - one size_t is far, far less accurate.    For a folder we can put the hash of the filenames in the 4 crc uint32s, and maybe set a high bit in the size field, to ensure they don't collide with the fingerprint of any actual file.

Reply:
Will change to as discussed IRL.